File: leetCode/tree/medium/quiz235/java/SolutionV1.java
package tree.medium.quiz235.java;

public class SolutionV1 {
    public static void main(String[] args) {
        TreeNode n4 = new TreeNode(4, new TreeNode(3), new TreeNode(5));
        TreeNode n2 = new TreeNode(2, new TreeNode(0), n4);
        TreeNode n8 = new TreeNode(8, new TreeNode(7), new TreeNode(9));
        TreeNode root = new TreeNode(6, n2, n8);
        TreeNode answer = new SolutionV1().lowestCommonAncestor(root, new TreeNode(2), new TreeNode(4));
        System.out.println(answer.val);
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        while (root != null) {
            if (p.val > root.val && q.val > root.val) { // 오른쪽 서브 트리
                root = root.right;
            } else if (p.val < root.val && q.val < root.val) { // 왼쪽 서브 트리
                root = root.left;
            } else {
                return root;
            }
        }
        return null;
    }
}

File: leetCode/tree/hard/quiz124/kotlin/SolutionV1.kt
package tree.hard.quiz124.kotlin

import tree.hard.quiz124.kotlin.SolutionV1.TreeNode

fun main() {
    val n9 = TreeNode(9)
    val n15 = TreeNode(15)
    val n7 = TreeNode(7)
    val n20 = TreeNode(20).apply {
        left = n15
        right = n7
    }
    val root = TreeNode(-10).apply {
        left = n9
        right = n20
    }

    SolutionV1().maxPathSum(root)
}

class SolutionV1 {

    private var max = 0

    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }

    fun maxPathSum(root: TreeNode?): Int {
        helper(root)
        return max
    }

    private fun helper(root: TreeNode?): Int {
        if (root == null) return 0

        val left = maxOf(helper(root.left), 0)
        val right = maxOf(helper(root.right), 0)

        max = maxOf(max, root.`val` + left + right)

        return root.`val` + maxOf(left, right)
    }
}
